qiskit/circuit/singleton.py

```diff
@@ -0,0 +1,355 @@
+
+There are several moving parts to tackle here.  The behaviour of having ``XGate()`` return some
+singleton object that is an (inexact) instance of :class:`.XGate` but *without* calling ``__init__``
+requires us to override :class:`type.__call__ <type>`.  This means that :class:`.XGate` must have a
+metaclass that defines ``__call__`` to return the singleton instance.
+
+Next, we need to ensure that there *is* a singleton instance for ``XGate()`` to return.  This can be
+done dynamically on each call (i.e. check if the instance exists and create it if not), but since
+we also want that instance to be very special, it's easier to hook in and create it during the
+definition of the ``XGate`` type object.  This also has the advantage that we do not need to make
+the singleton object pickleable; we only need to specify where to retrieve it from during the
+unpickle, because the creation of the base type object will recreate the singleton.
+
+We want the singleton instance to:
+
+* be immutable; it should reject all attempts to mutate itself.
+* have exactly the same state as an ``XGate()`` would have had if there was no singleton handling.
+
+We do this in a three-step procedure:
+
+1. Before creating any singletons, we separately define the overrides needed to make an
+   :class:`~.circuit.Instruction` and a :class:`.Gate` immutable.  This is
+   ``_SingletonInstructionOverrides`` and ``_SingletonGateOverrides``.
+
+2. While we are creating the ``XGate`` type object, we dynamically *also* create a subclass of it
+   that has the immutable overrides in its method-resolution order in the correct place. These
+   override the standard methods / properties that are defined on the mutable gate (we do not
+   attempt to override any cases where the type object we are creating has extra inplace methods).
+
+3. We can't instantiate this new subclass, because when it calls ``XGate.__init__``, it will attempt
+   to set some attributes, and these will be rejected by immutability.  Instead, we first create a
+   completely regular ``XGate`` instance, and then we dynamically change its type to the singleton
+   class, freezing it.
+
+We could do this entirely within the metaclass machinery, but that would require ``XGate`` to be
+defined as something like::
+
+  class XGate(Gate, metaclass=_SingletonMeta, overrides=_SingletonGateOverrides): ...
+
+which is super inconvenient (or we'd have to have ``_SingletonMeta`` do a bunch of fragile
+introspection).  Instead, we use the :class:`abc.ABC`/:class:`abc.ABCMeta` pattern of defining a
+concrete middle class (:class:`SingletonGate` in the :class:`.XGate` case) that sets the metaclass,
+selects the overrides to be applied, and has an :meth:`~object.__init_subclass__` that applies the
+singleton-subclass-creation steps above.  The overrides are in separate classes so that *mutable*
+:class:`.XGate` instances do not have them in their own MROs; doing this is easier to implement, but
```

Comment:
Just in case people don't know the acronym:

```suggestion
:class:`.XGate` instances do not have them in their own MROs (method resolution order); doing this is easier to implement, but
```

Reply:
Done (with line wrapping) in f3c6a43311.